src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs

```diff
@@ -90,48 +75,47 @@ internal AuthenticationOptions BaseOptions
             }
         }
 
-        private static void OnSendingHeaderCallback(object state)
+        private async Task OnResponseStartingCallback()
         {
-            AuthenticationHandler handler = (AuthenticationHandler)state;
-            handler.ApplyResponse();
+            //var handler = (AuthenticationHandler)state;
+            await FinishResponseOnce();
         }
 
-        protected virtual Task InitializeCoreAsync()
+        private async Task FinishResponseOnce()
         {
-            return Task.FromResult(0);
+            if (!_finishCalled)
+            {
+                _finishCalled = true;
+                await FinishResponseAsync();
+                await HandleAutomaticChallengeIfNeeded();
+            }
         }
 
         /// <summary>
-        /// Called once per request after Initialize and Invoke.
+        /// Hook that is called when the response about to be sent
         /// </summary>
-        /// <returns>async completion</returns>
-        internal async Task TeardownAsync()
+        /// <returns></returns>
+        protected virtual Task FinishResponseAsync()
         {
-            try
-            {
-                await ApplyResponseAsync();
-            }
-            catch (Exception)
+            return Task.FromResult(0);
+        }
+
+        private async Task HandleAutomaticChallengeIfNeeded()
+        {
+            if (!ChallengeCalled && BaseOptions.AutomaticAuthentication && Response.StatusCode == 401)
             {
-                try
-                {
-                    await TeardownCoreAsync();
-                }
-                catch (Exception)
-                {
-                    // Don't mask the original exception
-                }
-                UnregisterAuthenticationHandler();
-                throw;
+                await HandleUnauthorizedAsync(new ChallengeContext(BaseOptions.AuthenticationScheme));
             }
-
-            await TeardownCoreAsync();
-            UnregisterAuthenticationHandler();
         }
 
-        protected virtual Task TeardownCoreAsync()
+        /// <summary>
+        /// Called once per request after Initialize and Invoke.
```

Comment:
Is this comment current? It sounds like it got copied from somewhere else. This doesn't get called until unwind.